### PR TITLE
add build-and-run-integration-tests job

### PR DESCRIPTION
**Pull request by ManoelMilchev**
# Description

## What

Add integration tests setup, use cimg:go to allow building before running the integration tests, it is the minimum viable image for the requirements. it is pretty similar to cimg:base when it comes to size, both around 500 mb.

Tested with the following service [concept-exporter](https://github.com/Financial-Times/concept-exporter/compare/master...fix/UPPSF-6109-use-rob-int-tests)

## Why

[JIRA](https://financialtimes.atlassian.net/browse/UPPSF-6109)

## Scope and particulars of this PR (Please tick all that apply)

- [ ] Tech hygiene (dependency updating & other tech debt)
- [ ] Bug fix
- [x] Feature
- [ ] Documentation
- [ ] Breaking change
- [ ] Minor change (e.g. fixing a typo, adding config)

## DoD - Ensure all relevant tasks are completed before marking this PR as "Ready for review"

- [x] Test coverage is not significantly decreased
- [x] All PR checks have passed
- [x] Changes are deployed on dev before asking for review
- [x] Documentation remains up-to-date
    - [ ] OpenAPI definition file is updated
    - [ ] README file is updated
    - [ ] Documentation is updated in upp-docs and upp-public-docs
    - [ ] Architecture diagrams are updated

___
This Pull Request follows the rules described in our [Pull Requests Guide](https://github.com/Financial-Times/upp-docs/tree/master/guides/pr-guide)

### CODE REVIEW

File: src/executors/compose-executor.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+description: >
+    This executor is based on a machine executor with Docker and
+    Docker Compose pre-installed, and installs Golang.
+parameters:
+  golang-version:
+    type: string
+    default: "1.24.0"
```
Comment (tsvetko24) [inline src/executors/compose-executor.yml:7]: nitpick: better use 1.24.2, as it is already released. Refer to: https://go.dev/dl/

File: src/jobs/build-and-run-integration-tests.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    type: string
+    default: "v1.64.4"
+  golangci-config:
+    description: >
+      Location for golangci config file to be used (downloaded with wget)
+    type: string
+    default: ""
+  golang-version:
+    type: string
+    default: "1.24.0"
```
Comment (tsvetko24) [inline src/jobs/build-and-run-integration-tests.yml:22]: nitpick: same as above comment here: better use 1.24.2, as it is already released. Refer to: https://go.dev/dl/

File: src/executors/compose-executor.yml (hunk 1 above)
Comment (ManoelMilchev) [reply src/executors/compose-executor.yml]: Done!

File: src/jobs/build-and-run-integration-tests.yml (hunk 2 above)
Comment (ManoelMilchev) [reply src/jobs/build-and-run-integration-tests.yml]: Done!

File: src/jobs/build-and-run-integration-tests.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
+description: >
```
Comment (epavlova) [inline src/jobs/build-and-run-integration-tests.yml:1]: **suggestion:** I think there are a couple of points worth discussing online:
- This introduces a new job, which would require manual updates across all existing repositories to adopt it. Is this intentional, or could we support backward compatibility in some way?
- The job does not include unit tests, so consumers would likely need to run both this job and the existing one. However, this leads to repeated steps (e.g., build, lint) — is there a way to reduce duplication?
- We previously agreed that only unit tests should contribute to our Coveralls reporting, but here the integration tests are also included. If someone uses both `build-and-test` and this new job, we’ll end up with multiple Coveralls uploads, potentially skewing the results.

**Review by epavlova — CHANGES_REQUESTED** [review]

File: src/jobs/build-and-run-integration-tests.yml (hunk 3 above)
Comment (ManoelMilchev) [reply src/jobs/build-and-run-integration-tests.yml:1]: Hi Eli, thank you very much for the comment. I will try to answer it step by step.

This introduces a new job, which would require manual updates across all existing repositories to adopt it. Is this intentional, or could we support backward compatibility in some way? - It is a new job intentionally as not all repositories contain integration tests.

The job does not include unit tests, so consumers would likely need to run both this job and the existing one. However, this leads to repeated steps (e.g., build, lint) — is there a way to reduce duplication? - Tests without tags (unit tests in our case) are considered default and are always executed unless explicitly filtered out. For example our current build-and-tests job run a script doing the following 
```
#!/bin/bash
go test -tags=integration -mod=readonly -race -cover -coverprofile="${COVERAGE_REPORT_FOLDER}"/coverage.out ./...
```
 and this will still run the unit tests.
 
We previously agreed that only unit tests should contribute to our Coveralls reporting, but here the integration tests are also included. If someone uses both build-and-test and this new job, we’ll end up with multiple Coveralls uploads, potentially skewing the results. - This is true, currently we don't filter our which part of the code will be considered for coverage, I will give it a thought.

With all that said, maybe I can try to modify the current build-and-test job to include running docker-compose integration tests as last step after coverage has been reported from the unit tests, this way we won't have to modify each service separately and it should work out of the box. But it needs to be tested first.

Comment (ManoelMilchev) [reply src/jobs/build-and-run-integration-tests.yml]: Hi again Eli 👋 . I have reworked the tests to try and fit the requirements. Now we are able to trigger compose integration tests  [by providing the compose-executor when calling the build-and-test job](https://github.com/Financial-Times/concept-exporter/commit/3a7415fd3f7350ee8bf6aca1563726b943dc6f3b), w[e also keep the orb backwards compatible ](https://github.com/Financial-Times/cm-annotations-ontology/compare/main...fix/test-orb-compatibility)when we are using different/default executors. Additionally after the change coverage from the compose integration tests is not reported to coveralls as it is running separately. With that I think we cover the requirement, but please double check and let me know if you think we can improve the orb further!

File: src/jobs/build-and-test.yml [hunk 4]
```diff
@@ -40,6 +40,11 @@ steps:
       github-token: <<parameters.github-token>>
   - test:
       coverage-report-folder: <<parameters.coverage-report-folder>>
+  - setup_remote_docker:
```
Comment (epavlova) [inline src/jobs/build-and-test.yml:43]: **nitpick (non-blocking):** One more option for defining these additional steps is to make them conditional based on a new parameter in the `build-and-test` job. It will explicitly enable the new step so the user is not relying on the knowledge that certain executor enables the step - [CircleCI docs](https://circleci.com/docs/reusing-config/#defining-conditional-steps).
Simplified example:
```
jobs:
  conditional-example:
    parameters:
      run_optional_step:
        type: boolean
        default: false
    steps:
      - checkout
      - when:
          condition:
            equal: [ "true", "<< parameters.run_optional_step >>" ]
          steps:
            - run: echo "Running optional step"
```

File: src/jobs/build-and-test.yml [hunk 5]
```diff
@@ -40,6 +40,11 @@ steps:
...
+      docker_layer_caching: true
+  - run:
+      name: Run integration tests
```
Comment (epavlova) [inline src/jobs/build-and-test.yml:46]: **nitpick (non-blocking):** As both steps could do integration tests we can name this step `Run docker integration tests`.

**Review by epavlova — DISMISSED** [review]

File: src/jobs/build-and-run-integration-tests.yml (hunk 3 above)
Comment (epavlova) [reply src/jobs/build-and-run-integration-tests.yml]: Thanks for addressing this one.

File: src/jobs/build-and-test.yml (hunk 4 above)
Comment (ManoelMilchev) [reply src/jobs/build-and-test.yml]: Good point. I have added a boolean flag to take care of that. Additionally I tried to automatically modify it in case we are using compose-executor. But that didn't seem to work, as we cannot really refer parameter of type executor in the condition properly.

File: src/jobs/build-and-test.yml (hunk 5 above)
Comment (ManoelMilchev) [reply src/jobs/build-and-test.yml]: Done!

File: src/jobs/build-and-test.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (lokendersinghft) [reply src/jobs/build-and-test.yml]: Thanks Mani. Just noticed that Eli had already suggested it earlier.

**Review by lokendersinghft — APPROVED** [review]

File: src/jobs/build-and-test.yml (hunk 4 above)
Comment (epavlova) [reply src/jobs/build-and-test.yml]: Thanks for addressing this one.

**Review by epavlova — APPROVED** [review]
